mir-ci/mir_ci/robot_libraries/Screencopy.py

```diff
@@ -0,0 +1,145 @@
+from PIL import Image
+from robot.api import logger
+from robot.api.deco import keyword, library
+from RPA.Images import Images
+from RPA.recognition.templates import ImageNotFoundError
+
+
+@library(scope="GLOBAL")
+class Screencopy(ScreencopyTracker):
+    """
+    A Robot Framework library for capturing screenshots from the
+    Wayland display and performing image template matching.
+
+    The client connects to the display upon entering the first keyword,
+    and disconnects when the library goes out of scope.
+
+    If WAYLAND_DISPLAY is not defined, it defaults to 'wayland-0'.
+    """
+
+    ROBOT_LISTENER_API_VERSION = 3
+    TOLERANCE = 0.8
+
+    def __init__(self) -> None:
+        self.ROBOT_LIBRARY_LISTENER = self
+        display_name = os.environ.get("WAYLAND_DISPLAY", "wayland-0")
+        super().__init__(display_name)
+        self._rpa_images = Images()
+        self.last_frame_count = 0
+        self.screenshot = None
+
+    @keyword
+    async def match(self, template: str, timeout: int = 5) -> List[dict]:
+        """
+        Grab screenshots and compare until there's a match with the provided
+        template.
+
+        :param template: path to an image file to be used as template
+        :param timeout: timeout in seconds
+        :return: list of matched regions
+        :raises ImageNotFoundError: if no match is found within the timeout
+        """
+        regions = []
+        end_time = time.time() + float(timeout)
+        last_checked_frame_count = 0
+        while time.time() <= end_time:
+            self.screenshot = await self.grab_screenshot()
+            if last_checked_frame_count != self.frame_count:
+                last_checked_frame_count = self.frame_count
+                try:
+                    regions = self._rpa_images.find_template_in_image(
+                        self.screenshot,
+                        template,
+                        tolerance=self.TOLERANCE,
+                    )
+                except (RuntimeError, ValueError, ImageNotFoundError):
+                    continue
+                else:
+                    break
+        else:
+            self._log_failed_match(self.screenshot, template)
+            raise ImageNotFoundError
+
+        return [
+            {
+                "left": region.left,
+                "top": region.top,
+                "right": region.right,
+                "bottom": region.bottom,
+            }
+            for region in regions
+        ]
+
+    async def grab_screenshot(self):
+        """
+        Grabs the next frame tracked by the screencopy tracker.
+
+        :return Pillow Image of the next frame, or of the current
+            frame if the next frame is not available yet.
+        """
+        await self.connect()
+
+        # Wait for the first frame
+        while self.frame_count == 0:
+            await asyncio.sleep(0)
+
+        image = self.screenshot
+
+        if self.frame_count != self.last_frame_count:
```

Comment:
We don't want that check here now, do we?

Reply:
Oh, indeed

Reply:
Hmm, it now appears suspiciously similar to the original code of this PR. Am I missing something?